compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2595,8 +2604,10 @@ impl<'a> Parser<'a> {
             )?;
             let guard = if this.eat_keyword(kw::If) {
                 let if_span = this.prev_token.span;
+                this.is_in_a_valid_let_ctx = true;
```

Comment:
Use `with_let_expr_allowed` like I suggested above.

Reply:
Actually, you can probably use `with_let_expr_allowed` by moving it into `parse_expr` to fix the existing case of this code parsing (incorrectly):

```rust
if let a = 1 && { let x = let y = 1; } {}
```

Reply:
OK, it is more or less what `ast_validation.rs` already does.

> Actually, you can probably use `with_let_expr_allowed` by moving it into parse_expr

Just a random rant. The whole `expr.rs` file, as well as integrating `with_let_expr_allowed`, glues parsing logic with error management and this feature creates hard to understand and hard and maintain code.

Reply:
I certainly agree that `expr.rs` is a bit hard to follow.